### PR TITLE
fix: Localize Date and Time translations using AM/PM

### LINKED ISSUE #5491: Date and Time translations using AM/PM
Date and time are translated as they use AM/PM which is not used in other languages. Propose and implement a solution or translation across languages e.g. using 24h clock in other languages.

Screenshot from 2020-11-05 01-17-43

There are different time formats on the calendar page for example, e.g. https://eventyay.com/e/1f720482/calendar
Screenshot from 2020-11-05 12-28-11

### PR DESCRIPTION
**Pull request by daretobedifferent18**
Fixes #5491

#### Short description of what this resolves:
Time format is changed to 24hr clock whose languages are not supported by moment

#### Changes proposed in this pull request:

- 24 hr clock format is shown if AM/PM is not translated to resp language by moment

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by daretobedifferent18** [comment]
Wherever moment was able to convert it into resp language there we are using as it is otherwise 24hr clock.

Some ScreenShots
Screenshot from 2020-11-25 13-48-24 
Screenshot from 2020-11-25 13-48-46 
Screenshot from 2020-11-25 13-49-10
Screenshot from 2020-11-25 13-49-33
Screenshot from 2020-11-25 13-52-39
Screenshot from 2020-11-25 13-52-52
Screenshot from 2020-11-25 13-53-27
Screenshot from 2020-11-25 13-53-45
Screenshot from 2020-11-25 13-54-00
Screenshot from 2020-11-25 13-54-24
Screenshot from 2020-11-25 13-54-40
Screenshot from 2020-11-25 13-55-10 Screenshot from 2020-11-25 13-55-32

**Comment by daretobedifferent18** [comment]
Screenshot from 2020-11-25 14-07-36
Screenshot from 2020-11-25 14-07-54
Screenshot from 2020-11-25 14-08-08

**Comment by mariobehling** [comment]
Thank you. Could you change the following, please:
* no space in front of any commas
* change 0:00 to 24:00

**Comment by iamareebjamal** [comment]
We won't be able to change 0:00 to 24:00, that is controlled by moment JS AFAIK

**Comment by daretobedifferent18** [comment]
> We won't be able to change 0:00 to 24:00, that is controlled by moment JS AFAIK

exactly

File: .env.example [hunk 1]
```diff
@@ -1,4 +1,4 @@
-GOOGLE_API_KEY="Sample Key"
+GOOGLE_API_KEY="Sample Key "
```
Comment (iamareebjamal) [inline .env.example:1]: revert

File: app/helpers/general-date.js [hunk 2]
```diff
@@ -3,7 +3,19 @@ import moment from 'moment';
 
 export function generalDate(params) {
   const timezone = params[1] || moment.tz.guess();
-  const format = params[2] || 'h:mm A, MMMM Do YYYY (z)';
+  let changeTo24 = false;
+
+  let format = params[2] || 'h:mm A, MMMM Do YYYY (z)';
+
+  if (moment(params[0]).tz(timezone).format('A') === 'AM' || moment(params[0]).tz(timezone).format('A') === 'PM') {
+    changeTo24 = true;
```
Comment (iamareebjamal) [inline app/helpers/general-date.js:11]: I don't understand. What is this doing?

**Comment by iamareebjamal** [comment]
I think there will be conflicting changes in this PR and standardize PR, please work with @maze-runnar to finalize that PR and then work on this PR

File: app/helpers/general-date.js (hunk 2 above)
Comment (daretobedifferent18) [reply app/helpers/general-date.js:11]: I have checked for diff languages AM/PM is being converted by moment but not all the languages are supported in moment so the languages that are not supported gives AM/PM by default. So checking if they give AM/PM and if yes then converting format to 24 hr clock.

Comment (iamareebjamal) [reply app/helpers/general-date.js:11]: But in English, we want AM/PM

Comment (daretobedifferent18) [reply app/helpers/general-date.js:11]: for that I need to see what can I do

Comment (daretobedifferent18) [reply app/helpers/general-date.js:11]: is there a way in which I can get which language is currently active or to how to get which language moment is using?

Comment (iamareebjamal) [reply app/helpers/general-date.js:11]: It's better to maintain a set of locales for 12 hr/24 hr and switch based on detected locale

File: app/helpers/general-date.js [hunk 3]
```diff
@@ -3,7 +3,17 @@ import moment from 'moment';
 
 export function generalDate(params) {
   const timezone = params[1] || moment.tz.guess();
-  const format = params[2] || 'h:mm A, MMMM Do YYYY (z)';
+
+  const locals = ['en', 'bn', 'hi', 'id', 'ja', 'run', 'th', 'vi', 'ko'];
```
Comment (iamareebjamal) [inline app/helpers/general-date.js:7]: Make it a Set. Extract it outside the function
Change name to `locales12Hours`

**Comment by daretobedifferent18** [comment]
@iamareebjamal in many places moment-format is being used, should we enhance this translation there also, about AM/PM
bcz their also AM/PM is being used as a format and is resulting in AM/PM for all languages.
Should we use general-date for that also. and modify little bit to general-info to handle that cases as well??

**Comment by iamareebjamal** [comment]
Let's finalize date standardization first please

**Comment by daretobedifferent18** [comment]
> Let's finalize date standardization first please

ok after #5747 ?

**Comment by iamareebjamal** [comment]
Yes

**Review by NaviRocker — CHANGES_REQUESTED** [review]
So many Commits. Please Squash

**Comment by daretobedifferent18** [comment]
> @iamareebjamal in many places moment-format is being used, should we enhance this translation there also, about AM/PM
> bcz their also AM/PM is being used as a format and is resulting in AM/PM for all languages.
> Should we use general-date for that also. and modify little bit to general-info to handle that cases as well??

@iamareebjamal ?

**Comment by iamareebjamal** [comment]
Please list the areas where moment format is being used

File: app/helpers/header-date.js [hunk 4]
```diff
@@ -1,9 +1,21 @@
 import Helper from '@ember/component/helper';
 import moment from 'moment';
 
+const locales12Hours = new Set(['en', 'bn', 'hi', 'id', 'ja', 'run', 'th', 'vi', 'ko']);
+
 export function headerDate(params) {
```
Comment (iamareebjamal) [inline app/helpers/header-date.js:6]: If header date is not being used, please remove

Comment (daretobedifferent18) [reply app/helpers/header-date.js:6]: Used in various areaa
Screenshot from 2020-11-26 23-49-52

Comment (iamareebjamal) [reply app/helpers/header-date.js:6]: OK, please change these to general date as well

**Comment by daretobedifferent18** [comment]
> Please list the areas where moment format is being used

@iamareebjamal moment-format is used in :-
- access code/ Discount code
- attendee/invoices/billing/orders/messages/users/sales list
- Explore/Session/schedule route
- session card/ revenue card/footer main

**Comment by daretobedifferent18** [comment]
@iamareebjamal we can use some if conditions and handle everything inside general-date only and replace header-date and moment-format with general-date??

**Comment by iamareebjamal** [comment]
Yes

**Comment by daretobedifferent18** [comment]
@iamareebjamal I have replaced all the moment-format which are being rendered and header-date with general-date and it is working fine.
There are some places where the format is still not standardized.

**Comment by daretobedifferent18** [comment]
@iamareebjamal Pls see the changes, improved the format wherever not standardized, and also replaced everything with general-date.
Pls, let me know to make any changes further.

File: app/helpers/general-date.js [hunk 5]
```diff
@@ -8,7 +8,7 @@ export function generalDate(params) {
 
   const local = moment(params[0]).tz(timezone).locale();
 
-  let format = params[2] || 'h:mm A, MMMM Do YYYY (z)';
+  let format = params[2] || 'dddd, D MMMM, YYYY h:mm A (z)';
```
Comment (iamareebjamal) [inline app/helpers/general-date.js:11]: We can't change the default param just like that, it'll mess up the system elsewhere

Comment (daretobedifferent18) [reply app/helpers/general-date.js:11]: Ok I will revert it.

Comment (iamareebjamal) [reply app/helpers/general-date.js:11]: No

File: app/components/ui-table/cell/cell-date.hbs [hunk 6]
```diff
@@ -1 +1 @@
-{{general-date @record @props.options.timezone @props.options.dateFormat}}
+{{general-date @record @props.options.timezone '' @props.options.dateFormat}}
```
Comment (iamareebjamal) [inline app/components/ui-table/cell/cell-date.hbs:1]: Why the empty string?

Comment (daretobedifferent18) [reply app/components/ui-table/cell/cell-date.hbs:1]: sorry, not required here, will revert back.

**Comment by daretobedifferent18** [comment]
@iamareebjamal This is done now with all the changes said by you.

File: app/templates/components/event-invoice/event-info.hbs [hunk 7]
```diff
@@ -12,9 +12,9 @@
   <div class="ui padded segment">
     <strong>{{t 'At'}} {{this.event.locationName}}</strong>
     <br>
-    <strong>{{t 'From'}}:</strong> {{header-date this.event.startsAt}}
+    <strong>{{t 'From'}}:</strong> {{general-date this.event.startsAt}}
```
Comment (iamareebjamal) [inline app/templates/components/event-invoice/event-info.hbs:15]: I think header date used a different format than general-date?

Comment (daretobedifferent18) [reply app/templates/components/event-invoice/event-info.hbs:15]: what we can do is that wherever header-format is being used, we can manually pass format as param

Comment (daretobedifferent18) [reply app/templates/components/event-invoice/event-info.hbs:15]: it uses ```'dddd, MMMM Do YYYY, h:mm A'``` either we need to little modify this format and match it with the standardized formats or use this only.

Comment (iamareebjamal) [reply app/templates/components/event-invoice/event-info.hbs:15]: Exactly

Comment (daretobedifferent18) [reply app/templates/components/event-invoice/event-info.hbs:15]: use this only?

Comment (iamareebjamal) [reply app/templates/components/event-invoice/event-info.hbs:15]: match it with standard please

File: app/templates/components/event-invoice/event-info.hbs [hunk 8]
```diff
@@ -12,9 +12,9 @@
   <div class="ui padded segment">
     <strong>{{t 'At'}} {{this.event.locationName}}</strong>
     <br>
-    <strong>{{t 'From'}}:</strong> {{header-date this.event.startsAt}}
+    <strong>{{t 'From'}}:</strong> {{general-date this.event.startsAt '' 'dddd, D MMMM, YYYY h:mm A'}}
```
Comment (iamareebjamal) [inline app/templates/components/event-invoice/event-info.hbs:15]: I don't remember, was this the standard format which was chosen?

Comment (daretobedifferent18) [reply app/templates/components/event-invoice/event-info.hbs:15]: yes

Comment (daretobedifferent18) [reply app/templates/components/event-invoice/event-info.hbs:15]: What I am found in #5747 is that at some places still time format is not standardized

Comment (iamareebjamal) [reply app/templates/components/event-invoice/event-info.hbs:15]: Please create a Pr to fix that as well

**Pull request merged**